Testnet/NFTAuctionStore-Royalties/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,293 @@
+    {
+        EnsureNotPayable();
+
+        var tokenContract = Message.Sender;
+
+        Assert(State.IsContract(tokenContract), "The Caller is not a contract.");
+
+        var seller = fromAddress == Address.Zero ? operatorAddress : fromAddress;
+
+        var parameters = Serializer.ToStruct<AuctionParam>(data);
+
+        Assert(parameters.StartingPrice > 0, "Price should be higher than zero.");
+
+        Assert(Address == GetOwner(tokenContract, tokenId), "The store contract is not owner of the token.");
+
+        var auction = GetAuctionInfo(tokenContract, tokenId);
+
+        Assert(!auction.OnAuction, "The token is already on auction.");
+
+        auction = new AuctionInfo
+        {
+            Seller = seller,
+            EndBlock = checked(Block.Number + parameters.Duration),
+            StartingPrice = parameters.StartingPrice,
+            OnAuction = true
+        };
+
+        SetAuctionInfo(tokenContract, tokenId, auction);
+
+        Log(new AuctionStartedLog { Contract = tokenContract, TokenId = tokenId, EndBlock = auction.EndBlock, Seller = auction.Seller, startingPrice = auction.StartingPrice });
+
+        return true;
+    }
+
+    private void TransferToken(Address contract, UInt256 tokenId, Address from, Address to)
+    {
+        var result = Call(contract, 0, "TransferFrom", new object[] { from, to, tokenId });
+
+        Assert(result.Success, "The token transfer failed.");
+    }
+
+
+    private Address GetOwner(Address contract, UInt256 tokenId)
+    {
+        var result = Call(contract, 0, "OwnerOf", new object[] { tokenId });
+
+        Assert(result.Success && result.ReturnValue is Address, "GetOwner method call failed.");
+
+        return (Address)result.ReturnValue;
+    }
+
+    private void EnsureNotPayable() => Assert(Message.Value == 0, "The method is not payable.");
+
+    private bool SupportsRoyaltyInfo(Address contract)
+    {
+        var cached = GetSupportsRoyalty(contract);
+
+        if (!string.IsNullOrEmpty(cached))
+            return cached == bool.TrueString;
+
+        var result = Call(contract, 0, "SupportsInterface", new object[] { 6u /* IRoyaltyInfo */ });
+
+        var supported = result.Success && result.ReturnValue is bool value && value;
+
+        SetSupportsRoyalty(contract, supported);
+
+        return supported;
+    }
+
+    private RoyaltyInfo GetRoyaltyInfo(Address contract, UInt256 tokenId, ulong price)
+    {
+        var supportsRoyaltyInfo = SupportsRoyaltyInfo(contract);
+
+        if (!supportsRoyaltyInfo)
+            return new RoyaltyInfo { Recipient = Address.Zero, Amount = 0UL };
+
+        var royaltyCall = Call(contract, 0, "RoyaltyInfo", new object[] { tokenId, price });
+
+        Assert(royaltyCall.Success, "Get royalty info failed.");
+
+        var result = royaltyCall.ReturnValue as object[];
+        return new RoyaltyInfo { Recipient = (Address)result[0], Amount = (ulong)result[1] };
```

Comment:
Might be good to assert some extras here: 
- result.Length >= 2
- result[0] is an Address
- result[1] is a ulong